File: src/data/dtn.ts
import { useSyncExternalStore } from 'react'
import { hashBundlePayload } from './canonical'

export interface CustodyReceipt {
  custodianId: string
  custodianName: string
  timestamp: string
  location: string
  action: 'CREATED' | 'CUSTODY_ACQUIRED' | 'GATEWAY_UPLINKED' | 'ACKNOWLEDGED'
}

export interface ADUBundle {
  bundleId: string
  originNodeId: string
  originName: string
  householdName: string
  cellId: string
  emergencyType: string
  coordinates: { lat: number; lng: number }
  medicalSummary: string
  bloodGroup?: string
  conditions?: string
  medication?: string
  isBedridden?: boolean
  priority: 'P0_CRITICAL' | 'P1_HIGH' | 'P2_ROUTINE'
  integrityHash: string
  encryptedHash?: string
  hashAlgo?: string
  status: 'PENDING_LOCAL' | 'IN_TRANSIT' | 'DELIVERED_COMMAND'
  createdAt: string
  custodian?: { id: string; name: string; location: string }
  custodyReceipts: CustodyReceipt[]
  deliveredAt?: string
  tampered?: boolean
}

const STORAGE_KEY_DTN = 'sahayam_dtn_bundles_v2'

// Initial seed bundle for realistic demonstration if none exists yet
const SEED_BUNDLES: ADUBundle[] = []

function playEmergencyChime() {
  if (typeof window === 'undefined' || !window.AudioContext) return
  try {
    const ctx = new (window.AudioContext || (window as unknown as { webkitAudioContext: typeof AudioContext }).webkitAudioContext)()
    const now = ctx.currentTime
    const osc1 = ctx.createOscillator()
    const gain1 = ctx.createGain()
    osc1.type = 'sine'
    osc1.frequency.setValueAtTime(880, now) // A5
    osc1.frequency.exponentialRampToValueAtTime(1174.66, now + 0.15) // D6
    gain1.gain.setValueAtTime(0.15, now)
    gain1.gain.exponentialRampToValueAtTime(0.001, now + 0.35)
    osc1.connect(gain1)
    gain1.connect(ctx.destination)
    osc1.start(now)
    osc1.stop(now + 0.35)
  } catch {
    // AudioContext permission may be restricted until user gesture
  }
}

class DtnStore {
  private bundles: ADUBundle[] = []
  private listeners: Array<() => void> = []
  private broadcastChannel: BroadcastChannel | null = null
  private eventSource: EventSource | null = null
  private connectedToRelay = false
  private reconnectTimeout: ReturnType<typeof setTimeout> | null = null
  private reconnectAttempts = 0

  constructor() {
    this.init()
  }

  private init() {
    if (typeof window === 'undefined') return

    // 1. Load from localStorage
    try {
      const stored = localStorage.getItem(STORAGE_KEY_DTN)
      if (stored) {
        this.bundles = JSON.parse(stored)
      } else {
        this.bundles = SEED_BUNDLES
        this.saveLocally()
      }
    } catch {
      this.bundles = SEED_BUNDLES
    }

    // 2. Setup BroadcastChannel for instant local cross-tab / cross-window sync
    try {
      if ('BroadcastChannel' in window) {
        this.broadcastChannel = new BroadcastChannel('sahayam_dtn_mesh')
        this.broadcastChannel.onmessage = (event) => {
          const { type, bundle } = event.data || {}
          if (type === 'DEMO_RESET' || type === 'CLEAR_BUNDLES') {
            this.bundles = []
            this.saveLocally()
            this.notify()
            return
          }
          if (bundle) {
            this.upsertBundle(bundle, false)
            if (type === 'NEW_SOS') {
              playEmergencyChime()
            }
          }
        }
      }
    } catch {
      // ignore broadcast channel error
    }

    // 3. Connect to Vite Multi-Device SSE Relay
    this.connectSse()

    // 4. Initial fetch of active bundles from server relay
    this.fetchServerBundles()
  }

  private connectSse() {
    if (typeof window === 'undefined' || !window.EventSource) return

    try {
      if (this.eventSource) {
        this.eventSource.close()
      }

      this.eventSource = new EventSource('/api/dtn/events')

      this.eventSource.addEventListener('init', async (e: MessageEvent) => {
        this.connectedToRelay = true
        this.reconnectAttempts = 0
        try {
          const serverBundles = JSON.parse(e.data) as ADUBundle[]
          if (Array.isArray(serverBundles)) {
            if (serverBundles.length === 0) {
              this.bundles = []
              this.saveLocally()
            } else {
              for (const b of serverBundles) {
                const valid = await this.verifyBundle(b as unknown as Record<string, unknown>)
                if (!valid) b.tampered = true
                this.upsertBundle(b, false)
              }
            }
          }
        } catch {
          // ignore parse error
        }
        this.notify()
      })

      this.eventSource.addEventListener('new_bundle', async (e: MessageEvent) => {
        try {
          const bundle = JSON.parse(e.data) as ADUBundle
          const valid = await this.verifyBundle(bundle as unknown as Record<string, unknown>)
          if (!valid) bundle.tampered = true
          this.upsertBundle(bundle, true)
          playEmergencyChime()
        } catch {
          // ignore error
        }
      })

      this.eventSource.addEventListener('custody_updated', async (e: MessageEvent) => {
        try {
          const bundle = JSON.parse(e.data) as ADUBundle
          const valid = await this.verifyBundle(bundle as unknown as Record<string, unknown>)
          if (!valid) bundle.tampered = true
          this.upsertBundle(bundle, true)
          playEmergencyChime()
        } catch {
          // ignore error
        }
      })

      this.eventSource.addEventListener('bundle_delivered', async (e: MessageEvent) => {
        try {
          const bundle = JSON.parse(e.data) as ADUBundle
          const valid = await this.verifyBundle(bundle as unknown as Record<string, unknown>)
          if (!valid) bundle.tampered = true
          this.upsertBundle(bundle, true)
          playEmergencyChime()
        } catch {
          // ignore error
        }
      })

      this.eventSource.onerror = () => {
        this.connectedToRelay = false
        this.notify()
        this.scheduleReconnect()
      }
    } catch {
      this.connectedToRelay = false
      this.scheduleReconnect()
    }
  }

  private scheduleReconnect() {
    if (this.reconnectTimeout) return
    this.reconnectAttempts++
    const delay = Math.min(1000 * Math.pow(1.5, this.reconnectAttempts), 10_000)
    this.reconnectTimeout = setTimeout(() => {
      this.reconnectTimeout = null
      if (!this.connectedToRelay) {
        this.connectSse()
      }
    }, delay)
  }

  private async fetchServerBundles() {
    try {
      const res = await fetch('/api/dtn/bundles')
      if (res.ok) {
        const data = await res.json()
        if (Array.isArray(data) && data.length > 0) {
          for (const b of data) {
            const valid = await this.verifyBundle(b)
            if (!valid) b.tampered = true
            this.upsertBundle(b, false)
          }
        }
      }
    } catch {
      // server relay might be offline, relies on local storage & broadcast channel
    }
  }

  public async verifyBundle(bundle: Record<string, unknown>): Promise<boolean> {
    const hash = (bundle.integrityHash || bundle.encryptedHash) as string | undefined
    if (!hash || typeof hash !== 'string') return false
    const computed = await hashBundlePayload(bundle)
    return hash.toLowerCase() === computed.toLowerCase()
  }

  private upsertBundle(bundle: ADUBundle, shouldSave = true) {
    const existingIndex = this.bundles.findIndex(b => b.bundleId === bundle.bundleId)
    if (existingIndex >= 0) {
      this.bundles[existingIndex] = bundle
    } else {
      this.bundles = [bundle, ...this.bundles]
    }
    if (shouldSave) {
      this.saveLocally()
    }
    this.notify()
  }

  private saveLocally() {
    if (typeof window === 'undefined' || !window.localStorage) return
    try {
      localStorage.setItem(STORAGE_KEY_DTN, JSON.stringify(this.bundles))
    } catch {
      // storage quota or permission
    }
  }

  public subscribe(listener: () => void) {
    this.listeners.push(listener)
    return () => {
      this.listeners = this.listeners.filter(l => l !== listener)
    }
  }

  private notify() {
    this.listeners.forEach(cb => cb())
  }

  public getBundles(): ADUBundle[] {
    return this.bundles
  }

  public getBundleById(id: string): ADUBundle | undefined {
    return this.bundles.find(b => b.bundleId === id)
  }

  public isRelayConnected(): boolean {
    return this.connectedToRelay
  }

  // Publish a new SOS from a Civilian device
  public async publishSos(params: {
    originNodeId: string
    originName: string
    householdName: string
    cellId: string
    emergencyType: string
    coordinates?: { lat: number; lng: number }
    medicalSummary?: string
    bloodGroup?: string
    conditions?: string
    medication?: string
    isBedridden?: boolean
    priority?: 'P0_CRITICAL' | 'P1_HIGH' | 'P2_ROUTINE'
  }): Promise<ADUBundle> {
    const now = new Date().toISOString()
    const bundleId = `adu-${Date.now()}-${Math.random().toString(36).substring(2, 7)}`

    const preliminaryBundle: ADUBundle = {
      bundleId,
      originNodeId: params.originNodeId,
      originName: params.originName,
      householdName: params.householdName,
      cellId: params.cellId || 'WYD-07C',
      emergencyType: params.emergencyType,
      coordinates: params.coordinates || { lat: 11.552, lng: 76.102 },
      medicalSummary: params.medicalSummary || 'None reported',
      bloodGroup: params.bloodGroup || 'Unknown',
      conditions: params.conditions || 'None',
      medication: params.medication || 'None',
      isBedridden: !!params.isBedridden,
      priority: params.isBedridden || params.emergencyType.toLowerCase().includes('medical') || params.emergencyType.toLowerCase().includes('trapped')
        ? 'P0_CRITICAL'
        : 'P1_HIGH',
      integrityHash: '',
      hashAlgo: 'SHA-256',
      status: 'PENDING_LOCAL',
      createdAt: now,
      custodyReceipts: [
        {
          custodianId: params.originNodeId,
          custodianName: `${params.originName} (Victim Node)`,
          timestamp: now,
          location: `${params.householdName} · Cell ${params.cellId || 'WYD-07C'}`,
          action: 'CREATED',
        },
      ],
    }

    const calculatedHash = await hashBundlePayload(preliminaryBundle as unknown as Record<string, unknown>)
    const newBundle: ADUBundle = {
      ...preliminaryBundle,
      integrityHash: calculatedHash,
      encryptedHash: calculatedHash,
    }

    this.upsertBundle(newBundle, true)

    // Notify local tabs
    this.broadcastChannel?.postMessage({ type: 'NEW_SOS', bundle: newBundle })

    // Send to Vite relay for multi-device broadcast
    try {
      await fetch('/api/dtn/sos', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(newBundle),
      })
    } catch {
      // offline transmission queued
    }

    playEmergencyChime()
    return newBundle
  }

  // Volunteer takes courier custody
  public async acceptCustody(params: {
    bundleId: string
    custodianId: string
    custodianName: string
    location?: string
  }): Promise<ADUBundle | undefined> {
    const bundle = this.bundles.find(b => b.bundleId === params.bundleId)
    if (!bundle) return undefined

    // Verify bundle integrity prior to accepting physical custody
    const isValid = await this.verifyBundle(bundle as unknown as Record<string, unknown>)
    if (!isValid) {
      bundle.tampered = true
      this.saveLocally()
      this.notify()
      return undefined
    }

    const now = new Date().toISOString()
    bundle.status = 'IN_TRANSIT'
    bundle.custodian = {
      id: params.custodianId,
      name: params.custodianName,
      location: params.location || 'Mundakkai Trail · Sector 4',
    }
    bundle.custodyReceipts.push({
      custodianId: params.custodianId,
      custodianName: params.custodianName,
      timestamp: now,
      location: params.location || 'Mundakkai Trail · Sector 4',
      action: 'CUSTODY_ACQUIRED',
    })

    this.upsertBundle(bundle, true)
    this.broadcastChannel?.postMessage({ type: 'CUSTODY_UPDATED', bundle })

    try {
      await fetch('/api/dtn/custody', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          bundleId: params.bundleId,
          custodianId: params.custodianId,
          custodianName: params.custodianName,
          location: params.location,
        }),
      })
    } catch {
      // ignore relay error
    }

    return bundle
  }

  // Volunteer or transport courier reaches gateway and flushes to DEOC command
  public async uplinkToGateway(params: {
    bundleId: string
    gatewayName?: string
  }): Promise<ADUBundle | undefined> {
    const bundle = this.bundles.find(b => b.bundleId === params.bundleId)
    if (!bundle) return undefined

    // Verify bundle integrity prior to uplink delivery
    const isValid = await this.verifyBundle(bundle as unknown as Record<string, unknown>)
    if (!isValid) {
      bundle.tampered = true
      this.saveLocally()
      this.notify()
      return undefined
    }

    const now = new Date().toISOString()
    bundle.status = 'DELIVERED_COMMAND'
    bundle.deliveredAt = now
    bundle.custodyReceipts.push({
      custodianId: 'deoc-gateway-kalpetta',
      custodianName: params.gatewayName || 'St. Thomas HSS Relay Gateway',
      timestamp: now,
      location: 'Kalpetta District Emergency Ops Centre',
      action: 'GATEWAY_UPLINKED',
    })

    this.upsertBundle(bundle, true)
    this.broadcastChannel?.postMessage({ type: 'BUNDLE_DELIVERED', bundle })

    try {
      await fetch('/api/dtn/uplink', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          bundleId: params.bundleId,
          gatewayName: params.gatewayName,
        }),
      })
    } catch {
      // ignore relay error
    }

    return bundle
  }

  // Clear or reset demo data across local storage, tabs, and server
  public resetDemo() {
    this.bundles = []
    this.saveLocally()
    this.notify()
    if (typeof window !== 'undefined') {
      try {
        localStorage.setItem(STORAGE_KEY_DTN, '[]')
        localStorage.removeItem('sahayam_dtn_bundles')
        localStorage.removeItem('sahayam_dtn_bundles_v1')
      } catch {
        // ignore localStorage error
      }
      this.broadcastChannel?.postMessage({ type: 'DEMO_RESET' })
      fetch('/api/dtn/bundles', { method: 'DELETE' }).catch(() => {})
      const apiUrl = (typeof import.meta !== 'undefined' && import.meta.env?.VITE_API_URL) ? import.meta.env.VITE_API_URL.replace(/\/+$/, '') : ''
      if (apiUrl) {
        fetch(`${apiUrl}/api/dtn/bundles`, { method: 'DELETE' }).catch(() => {})
      }
    }
  }
}

export const dtnStore = new DtnStore()

export function useDtnMesh() {
  const bundles = useSyncExternalStore(
    cb => dtnStore.subscribe(cb),
    () => dtnStore.getBundles()
  )

  const isConnected = useSyncExternalStore(
    cb => dtnStore.subscribe(cb),
    () => dtnStore.isRelayConnected()
  )

  const pendingBundles = bundles.filter(b => b.status === 'PENDING_LOCAL')
  const inTransitBundles = bundles.filter(b => b.status === 'IN_TRANSIT')
  const deliveredBundles = bundles.filter(b => b.status === 'DELIVERED_COMMAND')

  return {
    bundles,
    pendingBundles,
    inTransitBundles,
    deliveredBundles,
    isRelayConnected: isConnected,
    publishSos: dtnStore.publishSos.bind(dtnStore),
    acceptCustody: dtnStore.acceptCustody.bind(dtnStore),
    uplinkToGateway: dtnStore.uplinkToGateway.bind(dtnStore),
    verifyBundle: dtnStore.verifyBundle.bind(dtnStore),
    resetDemo: dtnStore.resetDemo.bind(dtnStore),
  }
}

File: backend/app/routers/dtn.py
import asyncio
import hashlib
import json
from datetime import datetime, timezone
from typing import Any, Optional
from fastapi import APIRouter, HTTPException, Request, Response, status
from fastapi.responses import StreamingResponse
from pydantic import BaseModel, Field
from app.db import get_db
from app.sse import sse_manager

router = APIRouter(prefix="/api/dtn", tags=["dtn"])

IMMUTABLE_FIELDS = [
    "bundleId",
    "originNodeId",
    "originName",
    "householdName",
    "cellId",
    "emergencyType",
    "coordinates",
    "medicalSummary",
    "bloodGroup",
    "conditions",
    "medication",
    "isBedridden",
    "priority",
    "createdAt",
]


def canonical_hash_bundle(bundle: dict) -> str:
    payload = {k: bundle[k] for k in IMMUTABLE_FIELDS if k in bundle and bundle[k] is not None}
    canonical_str = json.dumps(payload, sort_keys=True, separators=(",", ":"))
    return hashlib.sha256(canonical_str.encode("utf-8")).hexdigest()


class CustodyPayload(BaseModel):
    bundleId: str
    custodianId: str
    custodianName: str
    location: Optional[str] = "Field Trail"


class UplinkPayload(BaseModel):
    bundleId: str
    gatewayName: Optional[str] = "DEOC Emergency Gateway"


def fetch_full_bundle(cur, bundle_id: str) -> Optional[dict]:
    cur.execute(
        """
        SELECT 
            b.bundle_id as "bundleId",
            b.origin_node_id as "originNodeId",
            b.origin_name as "originName",
            b.household_name as "householdName",
            b.cell_id as "cellId",
            b.emergency_type as "emergencyType",
            ST_Y(b.coordinates::geometry) as lat,
            ST_X(b.coordinates::geometry) as lng,
            b.medical_summary as "medicalSummary",
            b.blood_group as "bloodGroup",
            b.conditions,
            b.medication,
            b.is_bedridden as "isBedridden",
            b.priority,
            b.integrity_hash as "integrityHash",
            b.hash_algo as "hashAlgo",
            b.status,
            b.created_at as "createdAt",
            b.delivered_at as "deliveredAt",
            b.custodian
        FROM dtn_bundles b
        WHERE b.bundle_id = %s
        """,
        (bundle_id,),
    )
    row = cur.fetchone()
    if not row:
        return None

    cur.execute(
        """
        SELECT custodian_id as "custodianId", custodian_name as "custodianName",
               ts as "timestamp", location, action
        FROM custody_receipts
        WHERE bundle_id = %s
        ORDER BY ts ASC
        """,
        (bundle_id,),
    )
    receipts = cur.fetchall()

    return {
        "bundleId": row["bundleId"],
        "originNodeId": row["originNodeId"],
        "originName": row["originName"],
        "householdName": row["householdName"],
        "cellId": row["cellId"],
        "emergencyType": row["emergencyType"],
        "coordinates": {"lat": float(row["lat"]), "lng": float(row["lng"])} if row["lat"] and row["lng"] else {"lat": 11.552, "lng": 76.102},
        "medicalSummary": row["medicalSummary"],
        "bloodGroup": row["bloodGroup"],
        "conditions": row["conditions"],
        "medication": row["medication"],
        "isBedridden": row["isBedridden"],
        "priority": row["priority"],
        "integrityHash": row["integrityHash"],
        "encryptedHash": row["integrityHash"],
        "hashAlgo": row["hashAlgo"],
        "status": row["status"],
        "createdAt": row["createdAt"].isoformat() if row["createdAt"] else None,
        "deliveredAt": row["deliveredAt"].isoformat() if row["deliveredAt"] else None,
        "custodian": row["custodian"] if isinstance(row["custodian"], dict) else json.loads(row["custodian"]) if row["custodian"] else None,
        "custodyReceipts": [
            {
                **rc,
                "timestamp": rc["timestamp"].isoformat() if rc["timestamp"] else None,
            }
            for rc in receipts
        ],
    }


def fetch_all_bundles(cur) -> list[dict]:
    cur.execute("SELECT bundle_id FROM dtn_bundles ORDER BY created_at DESC")
    ids = [r["bundle_id"] for r in cur.fetchall()]
    return [fetch_full_bundle(cur, bid) for bid in ids if bid]


@router.get("/events")
async def sse_events(request: Request):
    async def event_generator():
        queue = sse_manager.subscribe()
        try:
            # Send initial bundles list
            with get_db() as conn:
                with conn.cursor() as cur:
                    initial_bundles = fetch_all_bundles(cur)
            yield f"event: init\ndata: {json.dumps(initial_bundles)}\n\n"

            while True:
                if await request.is_disconnected():
                    break
                try:
                    # Wait for message or heartbeat timeout (25 seconds)
                    msg = await asyncio.wait_for(queue.get(), timeout=25.0)
                    yield msg
                except asyncio.TimeoutError:
                    yield ": heartbeat\n\n"
        finally:
            sse_manager.unsubscribe(queue)

    return StreamingResponse(
        event_generator(),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache, no-transform",
            "Connection": "keep-alive",
            "Access-Control-Allow-Origin": "*",
        },
    )


@router.get("/bundles")
def get_bundles():
    with get_db() as conn:
        with conn.cursor() as cur:
            return fetch_all_bundles(cur)


@router.post("/sos", status_code=status.HTTP_201_CREATED)
async def publish_sos(payload: dict[str, Any]):
    bundle_id = payload.get("bundleId")
    origin_id = payload.get("originNodeId")
    cell_id = payload.get("cellId")
    emergency_type = payload.get("emergencyType")
    hash_val = payload.get("integrityHash") or payload.get("encryptedHash")

    if not bundle_id or not origin_id or not cell_id or not emergency_type or not hash_val:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail={"error": "INVALID_PAYLOAD", "message": "Missing required bundle fields"},
        )

    # Verify SHA-256 against canonical immutable payload
    expected_hash = canonical_hash_bundle(payload)
    if hash_val.lower() != expected_hash.lower():
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail={"error": "HASH_MISMATCH", "message": "Payload does not match SHA-256 integrity hash"},
        )

    coords = payload.get("coordinates") or {"lat": 11.552, "lng": 76.102}
    now = payload.get("createdAt") or datetime.now(timezone.utc).isoformat()

    with get_db() as conn:
        with conn.cursor() as cur:
            cur.execute(
                """
                INSERT INTO dtn_bundles (
                    bundle_id, origin_node_id, origin_name, household_name, cell_id,
                    emergency_type, coordinates, medical_summary, blood_group, conditions,
                    medication, is_bedridden, priority, integrity_hash, hash_algo, status,
                    created_at, custodian
                ) VALUES (
                    %s, %s, %s, %s, %s,
                    %s, ST_SetSRID(ST_MakePoint(%s, %s), 4326), %s, %s, %s,
                    %s, %s, %s, %s, %s, %s,
                    %s, %s
                )
                ON CONFLICT (bundle_id) DO NOTHING
                """,
                (
                    bundle_id,
                    origin_id,
                    payload.get("originName", "Unknown Civilian"),
                    payload.get("householdName", "Unknown Household"),
                    cell_id,
                    emergency_type,
                    coords.get("lng", 76.102),
                    coords.get("lat", 11.552),
                    payload.get("medicalSummary", "None reported"),
                    payload.get("bloodGroup", "Unknown"),
                    payload.get("conditions", "None"),
                    payload.get("medication", "None"),
                    payload.get("isBedridden", False),
                    payload.get("priority", "P1_HIGH"),
                    hash_val,
                    payload.get("hashAlgo", "SHA-256"),
                    payload.get("status", "PENDING_LOCAL"),
                    now,
                    json.dumps(payload.get("custodian")) if payload.get("custodian") else None,
                ),
            )

            # Insert initial creation receipt
            cur.execute(
                """
                INSERT INTO custody_receipts (bundle_id, custodian_id, custodian_name, ts, location, action)
                VALUES (%s, %s, %s, %s, %s, %s)
                """,
                (
                    bundle_id,
                    origin_id,
                    f"{payload.get('originName', 'Victim')} (Victim Node)",
                    now,
                    f"{payload.get('householdName', 'Household')} · Cell {cell_id}",
                    "CREATED",
                ),
            )
            bundle = fetch_full_bundle(cur, bundle_id)
        conn.commit()

    if bundle:
        await sse_manager.broadcast("new_bundle", bundle)

    return {"ok": True, "bundle": bundle}


@router.post("/custody")
async def accept_custody(payload: CustodyPayload):
    now = datetime.now(timezone.utc)
    custodian_dict = {
        "id": payload.custodianId,
        "name": payload.custodianName,
        "location": payload.location or "Field Trail",
    }

    with get_db() as conn:
        with conn.cursor() as cur:
            cur.execute("SELECT bundle_id FROM dtn_bundles WHERE bundle_id = %s", (payload.bundleId,))
            if not cur.fetchone():
                raise HTTPException(status_code=404, detail={"error": "NOT_FOUND", "message": "Bundle not found"})

            cur.execute(
                """
                UPDATE dtn_bundles 
                SET status = 'IN_TRANSIT', custodian = %s
                WHERE bundle_id = %s
                """,
                (json.dumps(custodian_dict), payload.bundleId),
            )

            cur.execute(
                """
                INSERT INTO custody_receipts (bundle_id, custodian_id, custodian_name, ts, location, action)
                VALUES (%s, %s, %s, %s, %s, %s)
                """,
                (
                    payload.bundleId,
                    payload.custodianId,
                    payload.custodianName,
                    now,
                    payload.location or "Field Trail",
                    "CUSTODY_ACQUIRED",
                ),
            )
            bundle = fetch_full_bundle(cur, payload.bundleId)
        conn.commit()

    if bundle:
        await sse_manager.broadcast("custody_updated", bundle)

    return {"ok": True, "bundle": bundle}


@router.post("/uplink")
async def uplink_bundle(payload: UplinkPayload):
    now = datetime.now(timezone.utc)
    with get_db() as conn:
        with conn.cursor() as cur:
            cur.execute("SELECT bundle_id FROM dtn_bundles WHERE bundle_id = %s", (payload.bundleId,))
            if not cur.fetchone():
                raise HTTPException(status_code=404, detail={"error": "NOT_FOUND", "message": "Bundle not found"})

            cur.execute(
                """
                UPDATE dtn_bundles 
                SET status = 'DELIVERED_COMMAND', delivered_at = %s
                WHERE bundle_id = %s
                """,
                (now, payload.bundleId),
            )

            cur.execute(
                """
                INSERT INTO custody_receipts (bundle_id, custodian_id, custodian_name, ts, location, action)
                VALUES (%s, %s, %s, %s, %s, %s)
                """,
                (
                    payload.bundleId,
                    "deoc-gateway-kalpetta",
                    payload.gatewayName or "St. Thomas HSS Relay Gateway",
                    now,
                    "Kalpetta District Emergency Ops Centre",
                    "GATEWAY_UPLINKED",
                ),
            )
            bundle = fetch_full_bundle(cur, payload.bundleId)
        conn.commit()

    if bundle:
        await sse_manager.broadcast("bundle_delivered", bundle)

    return {"ok": True, "bundle": bundle}


@router.delete("/bundles")
async def reset_bundles():
    with get_db() as conn:
        with conn.cursor() as cur:
            cur.execute("TRUNCATE TABLE custody_receipts, dtn_bundles CASCADE")
        conn.commit()
    await sse_manager.broadcast("init", [])
    return {"ok": True, "message": "All DTN bundles cleared"}
